staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,333 +17,233 @@ limitations under the License.
-		specs = append(specs, localSpecInfo)
-	}
-	if len(specs) == 0 {
-		return &spec.Swagger{}, nil
-	}
-	sortByPriority(specs)
-	for _, specInfo := range specs {
-		if specToReturn == nil {
-			specToReturn = &spec.Swagger{}
-			*specToReturn = *specInfo.spec
-			// Paths and Definitions are set by MergeSpecsIgnorePathConflict
-			specToReturn.Paths = nil
-			specToReturn.Definitions = nil
-		}
-		if err := aggregator.MergeSpecsIgnorePathConflict(specToReturn, specInfo.spec); err != nil {
-			return nil, err
-		}
-	}
-
-	return specToReturn, nil
-}
-
-// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
-func (s *specAggregator) updateOpenAPISpec() error {
-	if s.openAPIVersionedService == nil {
-		return nil
+func (s *specAggregator) addLocalSpec(name string, spec cached.Data[*spec.Swagger]) {
+	service := v1.APIService{}
+	service.Name = name
+	info := &openAPISpecInfo{
+		apiService: service,
 	}
-	specToServe, err := s.buildOpenAPISpec()
-	if err != nil {
-		return err
-	}
-	return s.openAPIVersionedService.UpdateSpec(specToServe)
+	info.spec.Replace(spec)
+	s.openAPISpecs[name] = info
 }
 
-// tryUpdatingServiceSpecs tries updating openAPISpecs map with specified specInfo, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
-	if specInfo == nil {
-		return fmt.Errorf("invalid input: specInfo must be non-nil")
-	}
-	_, updated := s.openAPISpecs[specInfo.apiService.Name]
-	origSpecInfo, existedBefore := s.openAPISpecs[specInfo.apiService.Name]
-	s.openAPISpecs[specInfo.apiService.Name] = specInfo
-
-	// Skip aggregation if OpenAPI spec didn't change
-	if existedBefore && origSpecInfo != nil && origSpecInfo.etag == specInfo.etag {
-		return nil
-	}
-	klog.V(2).Infof("Updating OpenAPI spec because %s is updated", specInfo.apiService.Name)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		reason := "add"
-		if updated {
-			reason = "update"
+// buildMergeSpecLocked creates a new cached mergeSpec from the list of cached.
+func (s *specAggregator) buildMergeSpecLocked() cached.Data[*spec.Swagger] {
+	caches := make(map[string]cached.Data[*spec.Swagger], len(s.openAPISpecs))
+	for name, specInfo := range s.openAPISpecs {
+		caches[name] = cached.NewTransformer[*spec.Swagger](func(result cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+			if result.Err != nil {
+				return result
+			}
+			// Make a copy of the data so that we don't modify the cache.
+			swagger := *result.Data
+			swagger.Definitions = handler.PruneDefaults(swagger.Definitions)
+			return cached.NewResultOK(&swagger, result.Etag)
+		}, &specInfo.spec)
+	}
+	return cached.NewMerger(func(results map[string]cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		s.mutex.Lock()
```

Comment:
I don't like how we lock inside a "locked function". This is technically deferred to when there is a request on /openapi/v2 but I'm wondering if we can make that more clear.

Reply:
Discussed offline, very happy that we're removing this lock.